File: Back/service/colocationService.go
package service

import (
	"Colibris/model"
	"gorm.io/gorm"
)

type ColocationService struct {
	db *gorm.DB
}

func NewColocationService(db *gorm.DB) ColocationService {
	return ColocationService{db: db}
}

func (s *ColocationService) CreateColocation(colocation *model.Colocation) error {
	colocation.ColocMembers = append(colocation.ColocMembers, model.ColocMember{
		UserID:       colocation.UserID,
		ColocationID: colocation.ID,
	})
	return s.db.Create(colocation).Error
}

func (s *ColocationService) GetColocationById(id int) (*model.Colocation, error) {
	var colocation model.Colocation
	result := s.db.Preload("ColocMembers").Where("id = ?", id).First(&colocation)
	return &colocation, result.Error
}

func (s *ColocationService) GetAllColocations(page int, pageSize int) ([]model.Colocation, int64, error) {
	var colocations []model.Colocation
	var total int64

	s.db.Model(&model.Colocation{}).Count(&total)
	if page == 0 || pageSize == 0 {
		result := s.db.Find(&colocations)
		return colocations, total, result.Error
	}

	offset := (page - 1) * pageSize
	result := s.db.Limit(pageSize).Offset(offset).Find(&colocations)
	return colocations, total, result.Error
}

func (s *ColocationService) GetAllUserColocations(userId int) ([]model.Colocation, error) {
	var colocations []model.Colocation
	result := s.db.Preload("ColocMembers").Find(&colocations)
	if result.Error != nil {
		return nil, result.Error
	}

	var filteredColocations []model.Colocation
	userIDUint := uint(userId)

	for _, colocation := range colocations {
		if colocation.UserID == userIDUint {
			filteredColocations = append(filteredColocations, colocation)
			continue
		}
		for _, member := range colocation.ColocMembers {
			if member.UserID == userIDUint {
				filteredColocations = append(filteredColocations, colocation)
				break
			}
		}
	}

	return filteredColocations, nil
}

func (s *ColocationService) UpdateColocation(id int, colocationUpdates map[string]interface{}) (*model.Colocation, error) {
	var colocation model.Colocation
	if err := s.db.Model(&colocation).Where("id = ?", id).Updates(colocationUpdates).Error; err != nil {
		return nil, err
	}
	if err := s.db.Where("id = ?", id).First(&colocation).Error; err != nil {
		return nil, err
	}
	return &colocation, nil
}

func (s *ColocationService) DeleteColocation(id int) error {
	us := NewUserService(s.db)
	errTaskAndVote := us.DeleteTasksAndVotes(uint(id))
	if errTaskAndVote != nil {
		return errTaskAndVote
	}

	errColocMember := us.DeleteColocMembers(uint(id))

	if errColocMember != nil {
		return errColocMember
	}

	return s.db.Where("id = ?", id).Delete(&model.Colocation{}).Error
}

func (s *ColocationService) SearchColocations(query string) ([]model.Colocation, error) {
	var colocations []model.Colocation

	result := s.db.Where("name LIKE ? OR description LIKE ?", "%"+query+"%", "%"+query+"%").Find(&colocations)
	return colocations, result.Error
}

File: Front/lib/services/colocation_service.dart
import 'dart:developer';

import 'package:dio/dio.dart';
import 'package:colibris/colocation/colocation.dart';
import 'package:colibris/utils/dio.dart';
import 'package:colibris/website/share/secure_storage.dart';

class ColocationService {
  final Dio _dio = dio;

  Future<ColocationResponse> getAllColocations(
      {int page = 1, int pageSize = 5}) async {
    var headers = await addHeader();
    try {
      final response = await _dio.get(
        '/colocations',
        queryParameters: {
          'page': page,
          'pageSize': pageSize,
        },
        options: Options(headers: headers),
      );

      if (response.statusCode == 200) {
        List<Colocation> colocations = (response.data['colocations'] as List)
            .map((colocation) => Colocation.fromJson(colocation))
            .toList();
        int total = response.data['total'];
        return ColocationResponse(colocations: colocations, total: total);
      } else {
        throw Exception('Failed to load colocations');
      }
    } on DioException catch (e) {
      print('Error: ${e.response?.statusCode}');
      print('Response data: ${e.response?.data}');
      log('Dio error!');
      log('Response status: ${e.response?.statusCode}');
      log('Response data: ${e.response?.data}');
      throw Exception('Failed to load colocations');
    }
  }

  Future<List<Colocation>> searchColocations({required String query}) async {
    var headers = await addHeader();
    try {
      final response = await _dio.get(
        '/colocations/search',
        queryParameters: {
          'query': query,
        },
        options: Options(headers: headers),
      );

      if (response.statusCode == 200) {
        List<Colocation> colocations = (response.data as List)
            .map((colocation) => Colocation.fromJson(colocation))
            .toList();
        return colocations;
      } else {
        throw Exception('Failed to search colocations');
      }
    } on DioException catch (e) {
      print('Error: ${e.response?.statusCode}');
      print('Response data: ${e.response?.data}');
      log('Dio error!');
      log('Response status: ${e.response?.statusCode}');
      log('Response data: ${e.response?.data}');
      throw Exception('Failed to search colocations');
    }
  }

  Future<void> createColocation({
    required String name,
    required String description,
    required bool isPermanent,
    required double latitude,
    required double longitude,
    required String location,
  }) async {
    var headers = await addHeader();
    var userData = await decodeToken();

    try {
      final response = await _dio.post(
        '/colocations',
        data: {
          'name': name,
          "userId": userData['user_id'],
          'description': description,
          'isPermanent': isPermanent,
          'latitude': latitude,
          'longitude': longitude,
          'location': location,
        },
        options: Options(headers: headers),
      );

      if (response.statusCode != 201) {
        throw Exception('Failed to create colocation');
      }
    } on DioException catch (e) {
      print('Error: ${e.response?.statusCode}');
      print('Response data: ${e.response?.data}');
      log('Dio error!');
      log('Response status: ${e.response?.statusCode}');
      log('Response data: ${e.response?.data}');
      throw Exception('Failed to create colocation');
    }
  }

  Future<void> updateColocation(
      int colocationId, Map<String, dynamic> colocationUpdates) async {
    var headers = await addHeader();
    try {
      final response = await _dio.patch(
        '/colocations/$colocationId',
        data: colocationUpdates,
        options: Options(headers: headers),
      );

      if (response.statusCode == 200) {
        log('Colocation updated successfully');
      } else {
        throw Exception('Failed to update colocation');
      }
    } on DioException catch (e) {
      print('Error: ${e.response?.statusCode}');
      print('Response data: ${e.response?.data}');
      log('Dio error!');
      log('Response status: ${e.response?.statusCode}');
      log('Response data: ${e.response?.data}');
      throw Exception('Failed to update colocation');
    }
  }

  Future<void> deleteColocation(int colocationId) async {
    var headers = await addHeader();
    try {
      final response = await _dio.delete(
        '/colocations/$colocationId',
        options: Options(headers: headers),
      );

      if (response.statusCode == 200) {
        log('Colocation deleted successfully');
      } else {
        throw Exception('Failed to delete colocation');
      }
    } on DioException catch (e) {
      print('Error: ${e.response?.statusCode}');
      print('Response data: ${e.response?.data}');
      log('Dio error!');
      log('Response status: ${e.response?.statusCode}');
      log('Response data: ${e.response?.data}');
      throw Exception('Failed to delete colocation');
    }
  }

  Future<Map<String, dynamic>> getColocationById(int colocationId) async {
    var headers = await addHeader();
    try {
      print('Fetching colocation data...');
      final response = await _dio.get(
        '/colocations/$colocationId',
        options: Options(headers: headers),
      );

      if (response.statusCode == 200) {
        return response.data;
      } else {
        throw Exception('Failed to load colocation data');
      }
    } on DioException catch (e) {
      print('Error: ${e.response?.statusCode}');
      print('Response data: ${e.response?.data}');
      print('Dio error!');
      print('Response status: ${e.response?.statusCode}');
      print('Response data: ${e.response?.data}');
      throw Exception('Failed to load colocation data');
    }
  }
}

class ColocationResponse {
  final List<Colocation> colocations;
  final int total;

  ColocationResponse({required this.colocations, required this.total});
}
